rsocket-core/src/test/java/io/rsocket/RSocketClientTerminationTest.java

```diff
@@ -0,0 +1,64 @@
+package io.rsocket;
+
+import io.rsocket.RSocketClientTest.ClientSocketRule;
+import io.rsocket.util.EmptyPayload;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.test.StepVerifier;
+
+import java.nio.channels.ClosedChannelException;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.function.Function;
+
+@RunWith(Parameterized.class)
+public class RSocketClientTerminationTest {
+
+  @Rule
+  public final ClientSocketRule rule = new ClientSocketRule();
+  private Function<RSocket, ? extends Publisher<?>> interaction;
+
+  public RSocketClientTerminationTest(Function<RSocket, ? extends Publisher<?>> interaction) {
+    this.interaction = interaction;
+  }
+
+  @Test
+  public void testCurrentStreamIsTerminatedOnConnectionClose() {
+    RSocketClient rSocket = rule.socket;
+
+    Mono.delay(Duration.ofSeconds(1))
+        .doOnNext(v -> rule.connection.dispose())
+        .subscribe();
+
+    StepVerifier.create(interaction.apply(rSocket))
+        .expectError(ClosedChannelException.class)
+        .verify(Duration.ofSeconds(5));
```

Comment:
Not specific to this PR: Are these tests "wall clock" slow?  Should we tag someway to run in CI, but not in normal developer builds?

Reply:
Yes, this will be great improvement for local builds - will do that in follow-up. Also I was thinking about VirtualTimeScheduler, but eventually preferred approach with no magic in It as this way tests look less fragile